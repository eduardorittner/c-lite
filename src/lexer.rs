#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token {
    pub kind: TokenKind,
    pub token: String,
    pub offset: usize,
}

use crate::parser::Precedence;
use miette::Result;
impl Token {
    pub fn infix_prec(&self) -> Result<Precedence> {
        match self.kind {
            TokenKind::Plus | TokenKind::Minus => Ok(Precedence::AddSub),
            _ => Err(miette::miette! {
            labels = vec![miette::LabeledSpan::at(self.offset..self.offset + self.token.len(), "here")],
            help = format!("Unexpected token: {self:?}"),
            "Not a valid infix operator token"}),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    Ident,
    Literal(LiteralKind),
    Whitespace,

    // Operators
    Plus,
    PlusEq,
    Minus,
    MinusEq,
    Slash,
    SlashEq,
    Star,
    StarEq,
    Percent,
    Eq,
    Not,
    And,
    Or,

    // Comparison Operators
    EqEq,
    NotEq,
    AndAnd,
    OrOr,
    GreaterEq,
    Greater,
    SmallerEq,
    Smaller,

    // Separators
    OpenParen,
    CloseParen,
    OpenBracket,  // `[`
    CloseBracket, // `]`
    OpenBrace,    // `{`
    CloseBrace,   // `}`
    Comma,        // `,`
    Dot,          // `.`
    Colon,        // `:`
    Semicolon,    // `;`

    // boolean
    True,
    False,

    // Control-flow
    For,
    If,
    Return,
    While,

    // Data types
    Enum,
    Struct,
    Union,
    TagUnion,
    Type,

    Eof,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LiteralKind {
    Hex,
    Decimal,
    Binary,
    String,
}

pub struct Lexer<'src> {
    src: &'src str,
    rest: &'src str,
    pub offset: usize,
    finished: bool,
}

impl<'src> Lexer<'src> {
    pub fn new(src: &'src str) -> Self {
        Self {
            src,
            rest: src,
            offset: 0,
            finished: false,
        }
    }
}

impl<'src> Iterator for Lexer<'src> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        if self.finished {
            return None;
        }
        loop {
            let mut chars = self.rest.chars();
            let char = match chars.next() {
                Some(c) => c,
                None => {
                    self.finished = true;
                    return Some(Token {
                        kind: Eof,
                        token: String::new(),
                        offset: self.offset,
                    });
                }
            };
            let char_str = &self.rest[..char.len_utf8()];
            let char_onwards = self.rest;
            let char_offset = self.offset;

            self.rest = chars.as_str();
            self.offset += char.len_utf8();

            let simple_token = move |kind: TokenKind| {
                Some(Token {
                    kind,
                    token: char_str.to_string(),
                    offset: char_offset,
                })
            };

            enum State {
                Ident,
                Whitespace,
                NumberLit,
                StringLit,
                IfEqElse(char, TokenKind, TokenKind),
            }
            use TokenKind::*;

            let state = match char {
                '.' => return simple_token(Dot),
                ',' => return simple_token(Comma),
                ';' => return simple_token(Semicolon),
                ':' => return simple_token(Colon),
                '(' => return simple_token(OpenParen),
                ')' => return simple_token(CloseParen),
                '[' => return simple_token(OpenBracket),
                ']' => return simple_token(CloseBracket),
                '{' => return simple_token(OpenBrace),
                '}' => return simple_token(CloseBrace),
                '%' => return simple_token(Percent),
                '+' => State::IfEqElse('=', PlusEq, Plus),
                '-' => State::IfEqElse('=', MinusEq, Minus),
                '*' => State::IfEqElse('=', StarEq, Star),
                '/' => State::IfEqElse('=', SlashEq, Slash),
                '=' => State::IfEqElse('=', EqEq, Eq),
                '!' => State::IfEqElse('=', NotEq, Not),
                '>' => State::IfEqElse('=', GreaterEq, Greater),
                '<' => State::IfEqElse('=', SmallerEq, Smaller),
                '&' => State::IfEqElse('&', AndAnd, And),
                '|' => State::IfEqElse('|', OrOr, Or),
                '"' => State::StringLit,
                '\'' => State::StringLit,
                '0'..='9' => State::NumberLit,
                'a'..='z' | 'A'..='Z' | '_' => State::Ident,
                c if c.is_whitespace() => State::Whitespace,
                c => panic!("Unexpected character during lexing: '{c}'"),
            };

            match state {
                State::Whitespace => {
                    let first_non_whitespace = char_onwards
                        .find(|c: char| !c.is_whitespace())
                        .unwrap_or(char_onwards.len());

                    let whitespace = &char_onwards[..first_non_whitespace];
                    let whitespace_offset = whitespace.len() - char.len_utf8();
                    let token = Some(Token {
                        kind: TokenKind::Whitespace,
                        token: char_onwards[..whitespace.len()].to_string(),
                        offset: char_offset,
                    });

                    self.offset += whitespace_offset;
                    self.rest = &self.rest[whitespace_offset..];

                    return token;
                }
                State::Ident => {
                    let first_non_ident = char_onwards
                        .find(|c| !matches!(c,'a'..='z'| 'A'..='Z'| '_'| '0'..='9'))
                        .unwrap_or(char_onwards.len());

                    let lit = &char_onwards[..first_non_ident];
                    let lit_offset = lit.len() - char.len_utf8();

                    let kind = match lit {
                        "for" => For,
                        "if" => If,
                        "return" => Return,
                        "while" => While,
                        "enum" => Enum,
                        "struct" => Struct,
                        "union" => Union,
                        "tagunion" => TagUnion,
                        "type" => Type,
                        "true" => True,
                        "false" => False,
                        _ => Ident,
                    };

                    let token = Some(Token {
                        kind,
                        token: char_onwards[..lit.len()].to_string(),
                        offset: char_offset,
                    });

                    self.offset += lit_offset;
                    self.rest = &self.rest[lit_offset..];
                    return token;
                }
                State::IfEqElse(char, yes, no) => {
                    if self.rest.starts_with(char) {
                        let token = char_onwards[..char_str.len() + char.len_utf8()].to_string();
                        self.offset += char.len_utf8();
                        self.rest = &self.rest[char.len_utf8()..];
                        return Some(Token {
                            kind: yes,
                            token,
                            offset: char_offset,
                        });
                    } else {
                        return Some(Token {
                            kind: no,
                            token: char_str.to_string(),
                            offset: char_offset,
                        });
                    }
                }
                State::NumberLit => {
                    let (prefix, lit_kind) = match char_onwards {
                        c if c.starts_with("0x") => (2, LiteralKind::Hex),
                        c if c.starts_with("0b") => (2, LiteralKind::Binary),
                        _ => (0, LiteralKind::Decimal),
                    };

                    let first_non_number = match lit_kind {
                        LiteralKind::Decimal => {
                            char_onwards[prefix..].find(|c| !matches!(c, '0'..='9' | '_' | '`'))
                        }
                        LiteralKind::Binary => {
                            char_onwards[prefix..].find(|c| !matches!(c, '0' | '1'))
                        }
                        LiteralKind::Hex => char_onwards[prefix..]
                            .find(|c| !matches!(c, '0'..='9' | 'a'..='f' | 'A'..='F' | '_' | '`')),
                        _ => unreachable!(),
                    };

                    let first_non_number = match first_non_number {
                        Some(n) => n + prefix,
                        None => char_onwards.len(),
                    };

                    let lit = &char_onwards[..first_non_number];
                    let lit_offset = lit.len() - char.len_utf8();
                    let token = Some(Token {
                        kind: TokenKind::Literal(lit_kind),
                        token: char_onwards[..lit.len()].to_string(),
                        offset: char_offset,
                    });
                    self.offset += lit_offset;
                    self.rest = &self.rest[lit_offset..];

                    return token;
                }
                State::StringLit => {
                    let closing_quote = char_onwards
                        .char_indices()
                        .find(|(i, c)| {
                            *i != 0
                                && *c == char
                                && (*i == 0 || char_onwards.chars().nth(i - 1).unwrap() != '\\')
                        })
                        .map(|(i, c)| i + c.len_utf8())
                        .unwrap_or(char_onwards.len());

                    let lit = &char_onwards[..closing_quote];

                    let lit_offset = lit.len() - char.len_utf8();
                    let token = Some(Token {
                        kind: TokenKind::Literal(LiteralKind::String),
                        token: char_onwards[..lit.len()].to_string(),
                        offset: char_offset,
                    });
                    self.offset += lit_offset;
                    self.rest = &self.rest[lit_offset..];
                    return token;
                }
            };
        }
    }
}

mod tests {
    use crate::lexer::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn single_char_tokens() {
        let source = "+ - * / = . , : ; ! & |";
        let lexer = Lexer::new(&source);
        let result: Vec<Token> = lexer.into_iter().collect();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn double_char_tokens() {
        let source = "+= -= *= /= == != <= >=";
        let lexer = Lexer::new(&source);
        let result: Vec<Token> = lexer.into_iter().collect();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn keyword() {
        let source = "if for while return struct enum union tagunion type";
        let lexer = Lexer::new(&source);
        let result: Vec<Token> = lexer.into_iter().collect();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn idents() {
        let source = "ident _ident ident_123";
        let lexer = Lexer::new(&source);
        let result: Vec<Token> = lexer.into_iter().collect();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn number_literals() {
        let source = "1234 0x9821 0b011 0xfe10 100`000`000_1";
        let lexer = Lexer::new(&source);
        let result: Vec<Token> = lexer.into_iter().collect();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn string_literals() {
        let source = "\"abc\" 'abc' \"abc'\" 'abc\"'";
        let lexer = Lexer::new(&source);
        let result: Vec<Token> = lexer.into_iter().collect();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn string_literal_escaping() {
        let source: String = ['"', '\\', '\"', '"'].iter().collect();
        let lexer = Lexer::new(&source);
        let result: Vec<Token> = lexer.into_iter().collect();
        assert_debug_snapshot!(result);
    }
}
