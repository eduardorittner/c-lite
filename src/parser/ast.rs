
use std::fmt::Write;

use crate::lexer::Token;

// Expressions are anything in the language that has an associated value
// or that performs some computation over other expressions
// Note that assignments such as "a = 1" are also (binary) expressions
#[derive(Debug, Clone)]
pub enum ExprKind {
    String(Token),
    Char(char),
    Number(u64),
    Bool(bool),
    Ident(Token),
    PostUnary {
        // a++ or a--
        op: Token,
        rhs: Box<ExprKind>,
    },
    Unary {
        // a++ or a--
        op: Token,
        rhs: Box<ExprKind>,
    },
    Binary {
        lhs: Box<ExprKind>,
        op: Token,
        rhs: Box<ExprKind>,
    },
    Postfix {
        lhs: Box<ExprKind>,
        op: Token,
    },
    Assign {
        lhs: Box<ExprKind>,
        op: Token,
        rhs: Box<ExprKind>,
    },
    CompoundAssign {
        lhs: Box<ExprKind>,
        op: Token,
        rhs: Box<ExprKind>,
    },
    Ternary {
        question: Box<ExprKind>,
        yes: Box<ExprKind>,
        no: Box<ExprKind>,
    },
    Comma {
        lhs: Box<ExprKind>,
        op: Token,
        rhs: Box<ExprKind>,
    },
}

// <decl> ::= <var-decl> | <type-decl> | <struct-decl>
#[derive(Debug, Clone)]
pub struct Decl {
    pub token: Token,
    pub kind: DeclKind,
}

#[derive(Debug, Clone)]
pub enum DeclKind {
    VarDecl {
        ident: Token,
        spec: Option<TypeSpec>,
        init: Init,
    },
    TypeDecl {
        oldtype: TypeSpec,
        newtype: Token,
    },
    StructDecl {
        name: Token,
        fields: Vec<Field>,
    },
}

#[derive(Debug, Clone)]
pub enum Init {
    Scalar(ExprKind),
    Aggregate(Vec<Box<Init>>),
}

impl PrettyPrint for Init {
    fn pretty_fmt(&self, depth: usize) -> String {
        match self {
            Init::Scalar(expr) => {
                format!("init:\n{}", expr.indent_fmt(depth + 1))
            }
            Init::Aggregate(inits) => {
                let mut result = String::from("init:\n");
                for init in inits {
                    let _ = write!(result, "{}, ", init.indent_fmt(depth + 1));
                }
                result
            }
        }
    }
}

impl std::fmt::Display for Init {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.indent_fmt(0))
    }
}

#[derive(Debug, Clone)]
pub struct TypeSpec {
    pub token: Token,
    pub kind: TypeSpecKind,
}

impl PrettyPrint for TypeSpec {
    fn pretty_fmt(&self, depth: usize) -> String {
        format!("type: {}", self.token.token)
    }
}

impl std::fmt::Display for TypeSpec {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.indent_fmt(0))
    }
}

#[derive(Debug, Clone)]
pub enum TypeSpecKind {
    Type,
    UserType,
    Pointer,
}

// TODO implement PrettyPrint for Field
#[derive(Debug, Clone)]
pub struct Field {
    pub name: Token,
    pub r#type: TypeSpec,
}

#[derive(Debug, Clone)]
pub enum StmtKind {
    Block {
        statements: Vec<StmtKind>,
    },
    Decl {
        declaration: Decl,
    },
    Expression {
        expression: ExprKind,
    },
    If {
        token: Token,
        cond: ExprKind,
        iftrue: Box<StmtKind>,
        else_token: Option<Token>,
        iffalse: Option<Box<StmtKind>>,
    },
    While {
        token: Token,
        cond: ExprKind,
        block: Box<StmtKind>,
    },
}

#[derive(Debug, Clone)]
pub enum ExternalDecl {
    Decl(Decl),
    FnDecl(FnDecl, Vec<StmtKind>),
}

#[derive(Debug, Clone)]
pub struct FnDecl {
    pub token: Token,
    pub name: Token,
    pub params: Vec<Param>,
    pub ret: Option<TypeSpec>,
}

#[derive(Debug, Clone)]
pub struct Param {
    pub spec: TypeSpec,
    pub name: Token,
}

pub trait PrettyPrint {
    fn pretty_fmt(&self, depth: usize) -> String;

    fn indent_fmt(&self, depth: usize) -> String {
        let indent = "|".repeat(depth);
        format!("{}{}", indent, self.pretty_fmt(depth))
    }
}

impl PrettyPrint for StmtKind {
    fn pretty_fmt(&self, depth: usize) -> String {
        match self {
            StmtKind::Block { ref statements } => {
                let mut result = String::from("Statements:\n");
                for stmt in statements {
                    let _ = write!(result, "{}\n", stmt.indent_fmt(depth + 1));
                }
                result
            }
            StmtKind::Decl { ref declaration } => {
                format!("{}", declaration.pretty_fmt(depth))
            }
            StmtKind::Expression { ref expression } => {
                format!("{}", expression.pretty_fmt(depth))
            }
            StmtKind::If {
                token: _,
                else_token: _,
                ref cond,
                ref iftrue,
                ref iffalse,
            } => {
                if let Some(iffalse) = iffalse {
                    format!(
                        "If\n{}\n{}then:\n{}{}else:\n{}",
                        cond.indent_fmt(depth + 1),
                        "|".repeat(depth + 2),
                        iftrue.indent_fmt(depth + 3),
                        "|".repeat(depth + 2),
                        iffalse.indent_fmt(depth + 3)
                    )
                } else {
                    format!(
                        "If\n{}\n{}then:\n{}",
                        cond.indent_fmt(depth + 1),
                        "|".repeat(depth + 2),
                        iftrue.indent_fmt(depth + 3)
                    )
                }
            }
            StmtKind::While {
                token: _,
                ref cond,
                ref block,
            } => format!("While {cond}:\n{block}"),
        }
    }
}

impl std::fmt::Display for StmtKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.indent_fmt(0))
    }
}

impl PrettyPrint for Decl {
    fn pretty_fmt(&self, depth: usize) -> String {
        match self.kind {
            DeclKind::VarDecl {
                ref ident,
                ref spec,
                ref init,
            } => {
                if let Some(spec) = spec {
                    format!(
                        "Var Declaration: '{}'\n{}\n{}",
                        ident.token,
                        spec.indent_fmt(depth + 1),
                        init.indent_fmt(depth + 1)
                    )
                } else {
                    format!(
                        "Var Declaration: '{}'\n{}",
                        ident.token,
                        init.indent_fmt(depth + 1)
                    )
                }
            }
            DeclKind::TypeDecl {
                oldtype: ref spec,
                ref newtype,
            } => {
                format!(
                    "Type Decl:\n{}\n{}newtype: {}",
                    spec.indent_fmt(depth + 1),
                    "|".repeat(depth + 1),
                    newtype.token
                )
            }
            DeclKind::StructDecl {
                ref name,
                ref fields,
            } => {
                format!(
                    "Struct Decl:\n{}{}, fields: {:?}",
                    "|".repeat(depth + 1),
                    name.token,
                    fields
                )
            }
        }
    }
}

impl std::fmt::Display for Decl {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.indent_fmt(0))
    }
}

impl PrettyPrint for ExprKind {
    fn pretty_fmt(&self, depth: usize) -> String {
        match self {
            ExprKind::String(token) => {
                format!("String: {}", token.token.clone())
            }
            ExprKind::Char(c) => format!("'{c}'"),
            ExprKind::Number(n) => format!("'{n}'"),
            ExprKind::Bool(b) => format!("'{b}'"),
            ExprKind::Ident(token) => format!("'{}'", token.token),
            ExprKind::Unary { op, rhs } => {
                format!("Unary: {}\n{}", op.kind, rhs.indent_fmt(depth + 1))
            }
            ExprKind::Binary { lhs, op, rhs } => format!(
                "Binary: {}\n{}\n{}",
                op.kind,
                lhs.indent_fmt(depth + 1),
                rhs.indent_fmt(depth + 1)
            ),
            ExprKind::Postfix { lhs, op } => {
                format!("Postfix: {}\n{}", op.kind, lhs.indent_fmt(depth + 1))
            }
            ExprKind::Assign { lhs, rhs, .. } => format!(
                "Assign:\n{}\n{}",
                lhs.indent_fmt(depth + 1),
                rhs.indent_fmt(depth + 1)
            ),
            ExprKind::CompoundAssign { lhs, op, rhs } => format!(
                "CompoundAssign: {}\n{}\n{}",
                op.kind,
                lhs.indent_fmt(depth + 1),
                rhs.indent_fmt(depth + 1)
            ),
            ExprKind::Ternary { question, yes, no } => format!(
                "Ternary:\n{}\n{}\n{}",
                question.indent_fmt(depth + 1),
                yes.indent_fmt(depth + 1),
                no.indent_fmt(depth + 1)
            ),
            ExprKind::Comma { lhs, op, rhs } => format!(
                "Comma: {}\n{}\n{}",
                op.kind,
                lhs.indent_fmt(depth + 1),
                rhs.indent_fmt(depth + 1)
            ),
            ExprKind::PostUnary { op, rhs } => {
                todo!("Post Unary: {}\n{}", op.kind, rhs.indent_fmt(depth + 1))
            }
        }
    }
}

impl std::fmt::Display for ExprKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.indent_fmt(0))
    }
}
