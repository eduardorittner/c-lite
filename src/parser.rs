use crate::lexer::{Lexer, LiteralKind, Token, TokenKind};
use miette::Result;

// Taken from [Gnu C Reference](https://www.gnu.org/software/gnu-c-manual/gnu-c-manual.html#Operator-Precedence)
#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
pub enum Precedence {
    // derive(Ord) makes the first variant the highest and so on
    Primary,
    Call, // (), [], .field
    Unary,
    MulDiv, // '*' | '/' | '%'
    AddSub, // '+' | '-'
    BitShift,
    Inequality, // '>' | '<' | '>=' | '<='
    Equality,   // '=' | '!='
    BitAnd,
    BitXor,
    BitOr,
    LogicAnd,
    LogicOr,
    Ternary,
    Assign,
    Comma,
    None,
}

mod ast {
    use crate::lexer::Token;

    enum Node {}

    #[derive(Debug, Clone)]
    pub enum ExprKind {
        String(Token),
        Char(char),
        Number(u64),
        Bool(bool),
        Ident(Token),
        Unary {
            op: Token,
            rhs: Box<ExprKind>,
        },
        Binary {
            lhs: Box<ExprKind>,
            op: Token,
            rhs: Box<ExprKind>,
        },
    }
}

// Returns a Some(token) only if it matches the pattern
// Useful for not having to compare on enums without declaring their fields
// e.g. TokenKind::Literal(_)
macro_rules! match_next {
    ($parser:expr, $expected:pat) => {{
        let matched = match $parser.peek() {
            Some(token) => matches!(token.kind, $expected),
            None => false,
        };
        if matched {
            $parser.advance()
        } else {
            None
        }
    }};
}

use ast::*;

pub struct Parser<'src> {
    src: &'src str,
    ast: Vec<ExprKind>,
    lexer: Lexer<'src>,
    peeked: Option<Token>,
}

impl<'src> Parser<'src> {
    pub fn new(src: &'src str) -> Self {
        Self {
            src,
            ast: Vec::new(),
            lexer: Lexer::new(src),
            peeked: None,
        }
    }

    fn advance(&mut self) -> Option<Token> {
        if let Some(token) = self.peeked.clone() {
            self.peeked = None;
            Some(token)
        } else {
            self.lexer.next()
        }
    }

    fn peek(&mut self) -> Option<Token> {
        if let Some(token) = self.peeked.clone() {
            Some(token)
        } else {
            self.peeked = self.lexer.next();
            self.peeked.clone()
        }
    }

    fn consume(&mut self, expected: TokenKind, msg: &str) -> Result<()> {
        self.expect(expected, msg).map(|_| ())
    }

    fn expect(&mut self, expected: TokenKind, msg: &str) -> Result<Token> {
        self.expect_where(|t| t == expected, msg)
    }

    fn expect_where(
        &mut self,
        mut check: impl FnMut(TokenKind) -> bool,
        msg: &str,
    ) -> Result<Token> {
        match self.advance() {
            Some(token) if check(token.kind) => {
                Ok(token)
            },
            Some(token) => {
                Err(miette::miette! {
                    labels = vec![miette::LabeledSpan::at(token.offset..token.offset + token.token.len(), "here")],
                    help = format!("Unexpected token: {token:?}"),
                    "{msg}"
            }.with_source_code(self.src.to_string()))},
            None => Err(miette::miette! {
                        labels = vec![miette::LabeledSpan::at(self.lexer.offset - 1..self.lexer.offset, "here")],
                        "{msg}"
                    }.with_source_code(self.src.to_string())),
        }
    }

    pub fn parse(&'src mut self) -> Result<ExprKind> {
        self.expr(Precedence::Assign)
    }

    pub fn expr(&mut self) -> Result<ExprKind> {
        todo!()
    }

    // Consumes an expression until it finds a operator with lower precedence
    // than the previous one.
    // e.g. when parsing "a + b * c" we parse a + (b * c) but when parsing
    // "a * b + c" we return (a * b) and let expr() parse the remaining "+"
    // expression
    pub fn inner_expr(&mut self, prec: Precedence) -> Result<ExprKind> {
        let token = self.peek().expect("Expected expression");

        // Parse infix expression
        let mut lhs = match token.kind {
            TokenKind::OpenParen => self.grouping(),
            _ => unreachable!(),
        }?;

        loop {
            let current_prec = self.peek().unwrap().infix_prec()?;
            if prec <= current_prec {
                break;
            }

            let token = self.advance().unwrap();
            lhs = match token.kind {
                TokenKind::Plus | TokenKind::Minus => self.binary(lhs)?,
                _ => unreachable!(),
            }
        }

        Ok(ExprKind::Binary {
            lhs: Box::new(lhs),
            op: token,
            rhs: Box::new(rhs),
        })
    }

    // prefix_expr
    pub fn prefix_expr(&mut self) -> Result<ExprKind> {
        let token = self.advance().unwrap();
        let expr = match token.kind {
            TokenKind::OpenParen => self.grouping(),
            _ => unreachable!(),
        }?;
        Ok(expr)
    }

    pub fn grouping(&mut self) -> Result<ExprKind> {
        let expr = self.expr(Precedence::Assign)?;
        self.expect(
            TokenKind::CloseParen,
            "Expected closing paren ')' after expression",
        )?;
        Ok(expr)
    }

    pub fn binary(&mut self, lhs: ExprKind) -> Result<ExprKind> {
        let token = self.advance().unwrap();
        let prec = match token.kind {
            TokenKind::Plus | TokenKind::Minus => Precedence::AddSub,
            TokenKind::Star | TokenKind::Slash | TokenKind::Percent => Precedence::MulDiv,
            // TODO add other cases
            _ => unreachable!(),
        };

        let expr = self.expr(prec)?;

        Ok(ExprKind::Binary {
            lhs: Box::new(lhs),
            op: token,
            rhs: Box::new(expr),
        })
    }

    pub fn unary(&mut self) -> Result<ExprKind> {
        let token = self.advance().unwrap();

        let expr = self.expr(Precedence::Unary)?;

        Ok(ExprKind::Unary {
            op: token,
            rhs: Box::new(expr),
        })
    }

    // <immutable> ::= ( <exp> ) | <call> | <constant>
    pub fn immutable(&mut self) -> Result<ExprKind> {
        let token = self.peek().unwrap();

        let expr = match token.kind {
            TokenKind::OpenParen => {
                self.advance();
                let expr = self.expr()?;
                
            },
            _ => unreachable!()
        }

        
        todo!()
    }

    // <constant> ::= NUMBER | CHAR | STRING | true | false
    pub fn constant(&mut self) -> Result<ExprKind> {
        let token = self
            .peek()
            .expect("Should have token when parsing constant");

        match token.kind {
            TokenKind::Literal(LiteralKind::Decimal) => self.decimal_number(),
            TokenKind::Literal(LiteralKind::Hex) => self.hex_number(),
            TokenKind::Literal(LiteralKind::Binary) => self.binary_number(),
            TokenKind::Literal(LiteralKind::String) => self.string(),
            TokenKind::True => Ok(ExprKind::Bool(true)),
            TokenKind::False => Ok(ExprKind::Bool(false)),
            _ => unreachable!(),
        }
    }

    pub fn string(&mut self) -> Result<ExprKind> {
        todo!()
    }

    fn decimal_number(&mut self) -> Result<ExprKind> {
        let token = self
            .advance()
            .expect("Should have token when parsing decimal literal");
        assert!(token.kind == TokenKind::Literal(LiteralKind::Decimal));

        let digits = token.token.replace("_", "").replace("`", "");
        if let Ok(n) = digits.parse::<u64>() {
            Ok(ExprKind::Number(n))
        } else {
            Err(miette::miette! {
                labels = vec![miette::LabeledSpan::at(token.offset..token.offset + token.token.len(), "here")],
                "Invalid numeric literal"
            }.with_source_code(self.src.to_string()))
        }
    }

    fn binary_number(&mut self) -> Result<ExprKind> {
        let token = self
            .advance()
            .expect("Should have token when parsing binary literal");
        assert!(token.kind == TokenKind::Literal(LiteralKind::Binary));

        let digits = token.token.strip_prefix("0b").unwrap();

        if let Ok(n) = u64::from_str_radix(&digits, 2) {
            Ok(ExprKind::Number(n))
        } else {
            Err(miette::miette! {
            labels =  vec![miette::LabeledSpan::at(token.offset..token.offset + token.token.len(), "here")],
            "Invalid binary literal"
            }.with_source_code(self.src.to_string()))
        }
    }

    fn hex_number(&mut self) -> Result<ExprKind> {
        let token = self
            .advance()
            .expect("Should have token when parsing hex literal");
        assert!(token.kind == TokenKind::Literal(LiteralKind::Hex));

        let digits = token.token.strip_prefix("0x").unwrap();

        if let Ok(n) = u64::from_str_radix(&digits, 16) {
            Ok(ExprKind::Number(n))
        } else {
            Err(miette::miette! {
            labels =  vec![miette::LabeledSpan::at(token.offset..token.offset + token.token.len(), "here")],
            "Invalid hexadecimal literal"
            }.with_source_code(self.src.to_string()))
        }
    }
}
