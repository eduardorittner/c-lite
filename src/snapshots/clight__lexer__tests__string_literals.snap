---
source: src/lexer.rs
expression: result
snapshot_kind: text
---
[
    Token {
        kind: Literal(
            String,
        ),
        token: "\"abc\"",
        offset: 0,
    },
    Token {
        kind: Whitespace,
        token: " ",
        offset: 5,
    },
    Token {
        kind: Literal(
            String,
        ),
        token: "'abc'",
        offset: 6,
    },
    Token {
        kind: Whitespace,
        token: " ",
        offset: 11,
    },
    Token {
        kind: Literal(
            String,
        ),
        token: "\"abc'\"",
        offset: 12,
    },
    Token {
        kind: Whitespace,
        token: " ",
        offset: 18,
    },
    Token {
        kind: Literal(
            String,
        ),
        token: "'abc\"'",
        offset: 19,
    },
    Token {
        kind: Eof,
        token: "",
        offset: 25,
    },
]
